hr_period/models/hr_payslip_run.py

```diff
@@ -0,0 +1,192 @@
+from openerp.exceptions import Warning as UserError
+
+from .hr_fiscal_year import get_schedules
+
+
+class HrPayslipRun(models.Model):
+    _inherit = 'hr.payslip.run'
+
+    name = fields.Char(
+        'Name',
+        required=True,
+        readonly=True,
+        states={'draft': [('readonly', False)]},
+        default=lambda obj: obj.env['ir.sequence']
+        .get('hr.payslip.run')
+    )
+    company_id = fields.Many2one(
+        'res.company',
+        'Company',
+        states={'close': [('readonly', True)]},
+        default=lambda obj: obj.env.user.company_id
+    )
+    hr_period_id = fields.Many2one(
+        'hr.period',
+        string='Period',
+        states={'close': [('readonly', True)]}
+    )
+    date_payment = fields.Date(
+        'Date of Payment',
+        states={'close': [('readonly', True)]}
+    )
+    schedule_pay = fields.Selection(
+        get_schedules,
+        'Scheduled Pay',
+        states={'close': [('readonly', True)]}
+    )
+
+    @api.multi
+    @api.constrains('hr_period_id', 'company_id')
+    def _check_period_company(self):
+        for run in self:
+            if run.hr_period_id:
+                if run.hr_period_id.company_id != run.company_id:
+                    raise UserError("The company on the selected period must "
+                                    "be the same as the company on the  "
+                                    "payslip batch.")
+        return True
+
+    @api.multi
+    @api.constrains('hr_period_id', 'schedule_pay')
+    def _check_period_schedule(self):
+        for run in self:
+            if run.hr_period_id:
+                if run.hr_period_id.schedule_pay != run.schedule_pay:
+                    raise UserError("The schedule on the selected period must "
+                                    "be the same as the schedule on the "
+                                    "payslip batch.")
+        return True
+
+    @api.model
+    def get_default_schedule(self, company_id):
+        company = self.env['res.company'].browse(company_id)
+
+        fy_obj = self.env['hr.fiscalyear']
+
+        fys = fy_obj.search([
+            ('state', '=', 'open'),
+            ('company_id', '=', company.id),
+        ])
+
+        return (
+            fys[0].schedule_pay
+            if len(fys) else 'monthly'
+        )
+
+    @api.onchange('company_id', 'schedule_pay')
+    @api.one
+    def onchange_company_id(self):
+        schedule_pay = self.schedule_pay or \
+            self.get_default_schedule(self.company_id.id)
+
+        if len(self.company_id) and schedule_pay:
+            period = self.env['hr.period'].get_next_period(self.company_id.id,
+                                                           schedule_pay,)
+            self.hr_period_id = period.id if period else False,
+
+    @api.onchange('hr_period_id')
+    def onchange_period_id(self):
+        if len(self.hr_period_id):
+            self.date_start = self.hr_period_id.date_start
```

Comment:
there is a lot of self.hr_period_id.

please add a variable like

period = self.hr_period_id